### PR TITLE
Vanilla javascript websocket with microphone example

### PR DESCRIPTION
**Pull request by iprovalo**
Vanilla javascript websocket client using AudioWorklet

### CODE REVIEW

**Comment by iprovalo** [comment]
This example is using a microphone with AudioWorklet interface of the Web Audio API to communicate with vosk server via a websocket.

**Comment by iprovalo** [comment]
Adding a buffer to the AudioWorkletProcessor to improve the quality of the data transfer and to speed up the recognition.

File: client-samples/javascript/data-conversion-processor.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    process(inputs, outputs, parameters) {
+        const inputData = inputs[0][0];
+
+        if (this.isBufferFull()) {
+            this.flush()
+        }
+
+        if (!inputData) return
```
Comment (zhurlik) [inline client-samples/javascript/data-conversion-processor.js:30]: Hi,

I have tried to run your solution, but I was not able to send an audio stream to the vosk server, cause the function `process()` always exits in this place. Probably I have missed something or it's still in development, do you have any instructions how to verify it, I used a simple http server ('live-server') for runing. If you want I can collect more details and share that with you.

Thanks,
Vlad

Comment (iprovalo) [reply client-samples/javascript/data-conversion-processor.js:30]: @zhurlik 

[Audio Worklet](https://developer.mozilla.org/en-US/docs/Web/API/Web_Audio_API/Using_AudioWorklet) has replaced a deprecated ScriptProcessorNode.   One possible issue is that this new AudioWorkletProcessor only runs over https (see [this](https://developer.mozilla.org/en-US/docs/Web/API/AudioWorkletNode) doc), so you will need to test it running on secure server.  It should work in local environment if a localhost vosk is running, but it won't work with a remote http server.  

I initially tested it with now deprecated ScriptProcessorNode (I just added it to the PR - voice_client_with_script_processor.js), so it is possible to still use ScriptProcessorNode instead, which will communicate over plain http.  In order to switch between the two approaches, change the index.html line 7.

Any errors in the javascript console when this code runs?

Thank you,

Ivan

Comment (zhurlik) [reply client-samples/javascript/data-conversion-processor.js:30]: @iprovalo Thanks for you answer. Seems to be getting better, at least I was able to see `AudioBuffer` being sent. However, I never received a response from **Vosk Server**. I plan to dive into this and try to figure out what's going wrong. In the meantime, I have written an example that works. 
```javascript
const webSocket = new WebSocket('ws://localhost:2700');
const AudioContext = window.AudioContext || window.webkitAudioContext;
const audioCtx = new AudioContext();
const TARGET_SAMPLE_RATE = 8000;

const floatTo16BitPCM = (output, offset, input) => {
    for (let i = 0; i < input.length; i++ , offset += 2) {
        const s = Math.max(-1, Math.min(1, input[i]));
        output.setInt16(offset, s < 0 ? s * 0x8000 : s * 0x7FFF, true);
    }
};

const encodeRAW = (samples) => {
    const buffer = new ArrayBuffer(samples.length * 2);
    const view = new DataView(buffer);
    floatTo16BitPCM(view, 0, samples);
    return view;
};

// https://stackoverflow.com/questions/27598270/resample-audio-buffer-from-44100-to-16000
const convertAndSend = async (recordedChunks) => {
    console.log('>> Converting audio...');

    // directly received by the audioprocess event from the microphone in the browser
    const arrayBuffer = await recordedChunks[0].arrayBuffer();
    const sourceAudioBuffer = await audioCtx.decodeAudioData(arrayBuffer);
    const offlineCtx = new OfflineAudioContext(sourceAudioBuffer.numberOfChannels,
        sourceAudioBuffer.duration * sourceAudioBuffer.numberOfChannels * TARGET_SAMPLE_RATE, TARGET_SAMPLE_RATE);
    const buffer = offlineCtx.createBuffer(sourceAudioBuffer.numberOfChannels, sourceAudioBuffer.length,
        sourceAudioBuffer.sampleRate);
    // Copy the source data into the offline AudioBuffer
    for (let channel = 0; channel < sourceAudioBuffer.numberOfChannels; channel++) {
        buffer.copyToChannel(sourceAudioBuffer.getChannelData(channel), channel);
    }

    // Play it from the beginning.
    const source = offlineCtx.createBufferSource();
    source.buffer = sourceAudioBuffer;
    source.connect(offlineCtx.destination);
    source.start(0);

    offlineCtx.oncomplete = (e) => {
        // use resampled.getChannelData(x) to get an Float32Array for channel x.
        const resampled = e.renderedBuffer;
        // use this float32array to send the samples to the server or whatever
        const leftFloat32Array = resampled.getChannelData(0);
        const data = encodeRAW(leftFloat32Array);
        sendToVosk(data);
    }
    offlineCtx.startRendering();
};

const sendToVosk = (data) => {
    console.log('>> Sending to Vosk Server...');
    webSocket.send(new Blob([data], {type: 'audio/raw'}));
    webSocket.send(new Blob([JSON.stringify({eof: 1})]));
    console.log('>> Audio has been sent');
}

webSocket.onopen = (event) => {
    console.log('>> Connected');
    navigator.mediaDevices.getUserMedia({audio: true, video: false})
        .then((stream) => {
            const mediaRecorder = new MediaRecorder(stream);
            const recordedChunks = [];

            mediaRecorder.addEventListener('dataavailable', (e) => {
                if (e.data.size > 0) {
                    recordedChunks.push(e.data);
                }
            });

            mediaRecorder.addEventListener('stop', async () => await convertAndSend(recordedChunks));

            console.log('>> Start recording...')
            mediaRecorder.start();

            setTimeout(() => {
                console.log('>> Stop recording...');
                mediaRecorder.stop();
            }, 3000);
        })
        .catch((err) => {
            /* handle the error */
            console.error(err);
        });
}

webSocket.onmessage = (event) => {
    console.log(">> Response:");
    console.log(event.data);
}

webSocket.onclose = (event) => {
    console.log(">> OnClose:" + JSON.stringify(event));
}

webSocket.onerror = (event) => {
    console.log(">> Error:" + JSON.stringify(event));
}
```


In your implementation I have seen some errors on **Vosk Server** side, I would like to try different ways of sending data via web socket.

Thanks,
Vlad

Comment (iprovalo) [reply client-samples/javascript/data-conversion-processor.js:30]: @zhurlik thank you!

**Comment by nshmyrev** [comment]
Amazing, huge thanks!

**Pull request merged**